documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc

```diff
@@ -2093,14 +2095,59 @@ public class VehicleMapperImpl extends VehicleMapper {
 ----
 ====
 
-Only methods with return type `void` may be annotated with `@BeforeMapping` or `@AfterMapping`. The methods may or may not have parameters.
-
 If the `@BeforeMapping` / `@AfterMapping` method has parameters, the method invocation is only generated if all parameters can be *assigned* by the source or target parameters of the mapping method:
 
 * A parameter annotated with `@MappingTarget` is populated with the target instance of the mapping.
 * A parameter annotated with `@TargetType` is populated with the target type of the mapping.
 * Any other parameter is populated with a source parameter of the mapping, whereas each source parameter is used once at most.
 
+If the before/after-mapping method has a return type other than `void`, it will be checked to match the target type of the mapping methods.
+Methods that have matching parameters, but whose return type does not match, will *not* be used. If the method returns a non-null value,
+this value will be returned from the mapping method.
+
```

Comment:
Hmm, Sjaak, I actually think the description given by Pascal is a bit easier. Because there actually isn't any preference in the selection - everything that possibly matches is used (with the order being stable). What's new is that *if* the method returns a non-void type, then that type has to be assignable to the return type of the mapping method. Perhaps the sentences in the doc above can be rephrased to not include so many negations, which I guess is what you're aming for?